### DISCUSSION TITLE
In-memory cache of Parquet data?

### DISCUSSION THREAD

**Discussion by yamin-oanda** [duckdb/duckdb#6948] [Q&A] [answered]
Hi, I am currently researching DuckDB for a use case storing and querying time series data. One of my questions is, how much does DuckDB cache in memory and are there are knobs to limit memory usage? More context: I have a service (Scala/JVM) and I am thinking of using DuckDB as its core calculation engine. The data will be read-only, stored in a Parquet partitioned structure. Simplified Parquet schema:

```sql
create table candle(
  time timestamp not null,
  base_curr text not null,
  quote_curr text not null,
  rate decimal not null,
  yyyymm text generated always as ( -- Partition by this column
    year(time) || lpad(month(time), 2, '0')
  )
);
```

I am thinking to query this using an in-memory DB with a view like: `create view candle_view as select * from 'candle/*/*.parquet'`. I've read about the [`enable_object_cache` pragma](https://duckdb.org/docs/sql/pragmas#enable_object_cache-disable_object_cache) and it looks useful because over time there will be many Parquet files so it's good to index their contents. Just wondering if DuckDB caches the actual data in memory though? E.g. similarly to how SQLite caches database pages in memory according to the [`cache_size` pragma](https://sqlite.org/pragma.html#pragma_cache_size).

Does DuckDB have a similar concept of database pages cached in memory?

**Comment by Mytherin** [comment, marked as answer]
DuckDB does cache its own tables in memory using a buffer manager similar to SQLite, but those do not apply to Parquet files. The only caching we have for Parquet files is the `enable_object_cache` which caches the metadata of the files.

**Reply by chriszrc** [reply]
Ok, so if I have a duckdb database with a single large table (~50gb), and I have enough RAM to put that in memory, how would I go about doing that?

**Reply by rodonn** [reply]
If you only need to read from the external Parquet table, one option is to explicitly create a new in memory table based on the Parquet input:
```
        CREATE OR REPLACE TABLE my_new_table_name AS
        SELECT *
        FROM read_parquet(my_external_table_path)
```

**Reply by pantonis** [reply]
@Mytherin Regarding ```enable_object_cache``` which caches the metadata of the files. 
Does this caching of the parquet files metadata happen only per session or they can be stored permantently on disk?

Thank you in advance

**Comment by yamin-oanda** [comment]
Got it, thank you. What about if I have a view which is a `union all` of a DuckDB table and a Parquet partition, to provide a complete view of my entire history? When I query the view, does DuckDB cache the part of the data that is loaded from its own table?

**Reply by Mytherin** [reply]
Yes, the DuckDB tables are always cached when memory is available.

**Reply by yamin-oanda** [reply]
Great, thanks again.

**Reply by kesavkolla** [reply]
I'm too greedy... Can metadata caching can live in different database which can be persisted to disk? I've large parquet files mostly read-only I want to cache metadata on to disk so that I don't have to fetch metadata all the time. If it can be in different database then I can attach that database and for querying I can use in memory database.
